src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationHttpHandler.Framework.cs

```diff
@@ -0,0 +1,51 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Security.Claims;
+using System.Web;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters.Authentication;
+
+/// <summary>
+/// HTTP handler for serving requests to remote authenticaiton endpoint.
+/// </summary>
+internal sealed class RemoteAuthenticationHttpHandler : IHttpHandler
+{
+    private readonly RemoteAuthenticationOptions _options;
+
+    public bool IsReusable => true;
+
+    public RemoteAuthenticationHttpHandler(RemoteAuthenticationOptions options)
+    {
+        if (string.IsNullOrEmpty(options.RemoteServiceOptions.ApiKey))
+        {
+            throw new ArgumentOutOfRangeException("API key must not be empty.");
+        }
+
+        _options = options;
+    }
+
+    public void ProcessRequest(HttpContext context)
+    {
+        if (context is null)
+        {
+            return;
+        }
+
+        // If a user is logged in (using ASP.NET's usual authenticaiton mechanisms), return that claims principal.
+        if (context.User is ClaimsPrincipal claimsPrincipal && context.User.Identity.IsAuthenticated)
+        {
+            context.Response.StatusCode = 200;
+            context.Response.ContentType = "application/octet-stream";
+            claimsPrincipal.WriteTo(new BinaryWriter(context.Response.OutputStream));
+        }
+        else
+        {
+            context.Response.StatusCode = 401;
```

Comment:
How do you expect the module to set challenges like 301 with a location header? Or 401 with a www-authenticate header?

Reply:
At least for the OWIN scenarios I've tried so far, the OWIN middleware for cookie auth or bearer auth actually takes care of that. Those components just look for a 401 response (which I set here) and update headers, status code, etc. as needed. I still need to confirm that this will also "magically work" for forms auth scenarios but will look into that this week.

Reply:
I've confirmed that this works for forms auth, too, so at least for most common kinds of auth the 401 returned by the handler is translated into the appropriate status code and headers.